detectron2/engine/defaults.py

```diff
@@ -390,6 +390,9 @@ def __init__(self, cfg):
             optimizer=optimizer,
             trainer=weakref.proxy(self),
         )
+        if cfg.SOLVER.AMP.ENABLED:
+            self.checkpointer.add_checkpointable('grad_scaler', self._trainer.grad_scaler)
+            
```

Comment:
To solve checkpointing issues for both optimizer and grad_scaler, I think it's better to implement a correct `state_dict()` for `DefaultTrainer`.

It can be implemented by merging `super().state_dict()` and `self._trainer.state_dict()`.